dijkstra/lib/graph.js

```diff
@@ -15,20 +29,16 @@ function generate(numNodes) {
   for (var i=0; i < numNodes; i++) {
     var node = [];
 
-    // assure every node is connected to at least one other by making a circle
+    // make circular edges
+    //  - assures every node is reachable from every other node
+    //  - assures every node has at least one incoming and one outgoing edge
     node.push(makeEdge(i === numNodes - 1 ? 0 : i + 1));
 
-    for (var e=0; e < getRandomInt(0, 4); e++) {
+    // make random edges
+    const edges = getRandomInt(0, MAX_RANDOM_EDGES);
+    for (var e=0; e < edges; e++) {
       var dest = getRandomInt(0, numNodes - 1);
-      if (
-        // avoid self-reference
-        dest !== i
-        &&
-        // avoid duplicate edges
-        node.find((e) => e.destination === dest) === undefined
-      ) {
-        node.push(makeEdge(dest));
-      }
+      if (isValidEdge(node, dest, i)) node.push(makeEdge(dest));
```

Comment:
Nit: if both the condition and the body contain some logic, it might be more convenient to have them on two separate lines — it makes debugging easier because you can step over each separately.